core/sched.cc

```diff
@@ -380,7 +405,10 @@ void cpu::reschedule_from_interrupt(bool called_from_yield,
         n->_runtime.add_context_switch_penalty();
     }
     preemption_timer.cancel();
-    if (n->_realtime._priority == 0) {
+    if (n->_realtime.has_slice()) {
+        assert(n->_realtime.has_remaining());
+        preemption_timer.set_with_irq_disabled(now + n->_realtime.remaining());
```

Comment:
Please make sure (I think I understood it, but not 100% sure) that the logic above gurantees that if !has_remaining we don't reach here at all.

In particular, what happens if some thread is the only thread in its priority group, so it continues to run? Does the code above recognize this case and remember to reset its time slice?

Reply:
> Please make sure (I think I understood it, but not 100% sure) that the logic above gurantees that if !has_remaining we don't reach here at all.

Unfortunately, I of cause don't have a proof that the assertion always holds. However, from understanding of the scheduler code, I believe that it should. My thinking was to add it and if we discover a path where it fails during testing than we might have found a bug. I can also remove the assertion if you want me to.

> In particular, what happens if some thread is the only thread in its priority group, so it continues to run? Does the code above recognize this case and remember to reset its time slice?

Shouldn't this case be covered by the `priority_precedence` test that I added? If not we should expand the test suite to cover this.